samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -281,14 +286,20 @@ public void onContainerFailed(Throwable t) {
               LOGGER.warn("JobCoordinatorLatch was null. It is possible for some component to be waiting.");
             }
             containerException = t;
+            SamzaContainer container = containerRef.getAndSet(null);
             LOGGER.error(String.format("Container: %s failed with an exception. Stopping the stream processor: %s. Original exception:", container, processorId), containerException);
-            container = null;
             stop();
           }
         };
 
-        container = createSamzaContainer(processorId, jobModel);
+        SamzaContainer container = createSamzaContainer(processorId, jobModel);
         container.setContainerListener(containerListener);
+
+        SamzaContainer previousContainer = containerRef.getAndSet(container);
+        if (previousContainer != null && previousContainer.getStatus().equals(SamzaContainerStatus.STARTED)) {
```

Comment:
Don't think this should be necessary if you update state atomically. It looks like you would not have the state for `previousContainer` in the first place?

Reply:
@vjagadish1989 
Agreed to this suggestion and switched to using explicit locks.